### PR TITLE
ipns address

### LINKED ISSUE #9: IPNS keygen support
🙏 it'd be nice to have ed25519 IPNS (libp2p-key) keygen support.

This is our experimental app-specific deterministic keygen format using deterministic ETH signatures for ENS contenthash. 
```
let caip10 = `eip155:1:${App.user.address}`
let domain = "domain.eth"
let info = `${caip10}:${domain}`
let password = "pass12#$" // optional salt/pin
let extradata = await sha256(`${info}:${password ? password : ''}`) //still testing 
let msg = `Requesting Signature To Generate IPNS Keys For ${domain}\n\nWARNING:Do Not Sign This Request From Untrusted Clients\nExtradata: ${extradata}\nSigned By: ${caip10}`;
let sig = await App.user.signMessage(msg);
let inputKey = sha256(
    hexToBytes(
        sig.toLowerCase().startsWith('0x') ? sig.slice(2) : sig
    )
)
let salt = await sha256(`${info}:${password ? password : ''}:${sig.slice(-64)}`)
let hashKey = await hkdf(sha256, inputKey, salt, info, 42)
let privateKey = hashToPrivateScalar(hashKey, ed25519.CURVE.n, true).toString(16).padStart(64, "0")
let publicKey = bytesToHex(await ed25519.getPublicKey(privateKey))
let key = `08011240${privateKey}${publicKey}`
let w3Name = await Name.from(hexToBytes(key))
// let contenthash = `0xe5010172002408011220{$publicKey}`

```

### PR DESCRIPTION
**Pull request by 0xc0de4c0ffee**
resolves #9  

- missing base36 support
- keygen only, missing IPNS records signing as it's more complicated using cbor.

I've some questions commented out on codes for review 🙏

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -15,11 +15,12 @@ For the apps made with the library, check out:
 
 > npm install ed25519-keygen
 
-The package exports five modules:
+The package exports six modules:
 
 - [`ed25519-keygen/ssh`](#sshseed-username) for SSH key generation
 - [`ed25519-keygen/pgp`](#pgpseed-user-password) for [RFC 4880](https://datatracker.ietf.org/doc/html/rfc4880) + [RFC 6637](https://datatracker.ietf.org/doc/html/rfc6637)
 - [`ed25519-keygen/tor`](#torseed) for TOR onion addresses
+- [`ed25519-keygen/ipns`](#ipnsseed) for IPNS addresses
```
Comment (0xc0de4c0ffee) [inline README.md:23]: IPNS (ed25519/libp2p-key)

File: src/ipns.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  // 0x01 = v1, 0x72 = libp2p-key, ?0x00240801 = ?ed25519 key/type, 0x12 = ?sha256, 0x20 = 32 bytes length
+  const pubKeyBytes = concatBytes([0x01, 0x72, 0x00, 0x24, 0x08, 0x01, 0x12, 0x20], pubKey);
+  const b16 = hex.encode(pubKeyBytes).toLowerCase();
+  return {
+    publicKey: `0x${b16}`,
+    privateKey: `0x${hex.encode(concatBytes([0x08, 0x01, 0x12, 0x40], seed, pubKey))}`,
+    //b58/ old format, not recommended
+    //base58: `${base58.encode(pubKeyBytes)}`,
+    //b36/ subdomain safe,recommended, ??scure/base36?
+    //base36: `k${base36.encode(pubKeyBytes).toLowerCase()}`,
```
Comment (0xc0de4c0ffee) [inline src/ipns.ts]: base36: k51qzi5uqu5dgnfwbc46une4upw1vc9hxznymyeykmg6rev1513yrnbyrwmmql

**Comment by paulmillr** [comment]
You can implement base36 by yourself. It should be straightforward, maybe 10 lines of code. Check out source code of scure-base for base32 implementation. It makes it really easy.

**Comment by paulmillr** [comment]
That base36 implementation doesn't seem right.

See this: https://github.com/paulmillr/scure-base/blob/b266c73dde977b1dd7ef40ef7a23cc15aab526b3/index.ts#L283

**Comment by 0xc0de4c0ffee** [comment]
Check this CID explorer

https://cid.ipfs.tech/#k51qzi5uqu5dhhcu1pop9pynjg2g3l6vrlt379x6huzy2zhyg54o1u6csnuwi3
https://cid.ipfs.tech/#bafzaajaiaejcaewifgpmfri57654wt47zsw45ykcjszdp2ntbu6nolkhygaqg2e5

**Comment by 0xc0de4c0ffee** [comment]
> That base36 implementation doesn't seem right.
> 
> See this: https://github.com/paulmillr/scure-base/blob/b266c73dde977b1dd7ef40ef7a23cc15aab526b3/index.ts#L283

I tried `scure-base` but I'm not sure about radix2 and padding for base36?
> base36	 k	active	base36 [0-9a-z] case-insensitive - no padding 
https://www.ietf.org/archive/id/draft-multiformats-multibase-04.html

`const base36 = utils.chain(utils.radix2(<??>), utils.alphabet('0123456789abcdefghijklmnopqrstuvwxyz'), utils.padding(<??>), utils.join(''));`

// throw new Error('radix2: bits should be in (0..32]'); 
// convertRadix2: invalid data word=35 from=5

**Comment by paulmillr** [comment]
it's not radix2 -- it's radix

for padding maybe no padding

File: src/ipns.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  if (hexKey.startsWith('0172002408011220') && hexKey.length === 80) {
+    return hex.decode(hexKey);
+  }
+  // Throw error if IPNS key prefix is invalid
+  throw new Error('Invalid IPNS Key Prefix: ' + hexKey);
+}
+
+// Generates an ed25519 pubkey from a seed and converts it to several IPNS pubkey formats
+export async function getKeys(seed: Uint8Array) {
+  //? privKey "seed" should be checked for <ed25519.curve.n?
```
Comment (0xc0de4c0ffee) [inline src/ipns.ts:44]: //? privKey "seed" should be checked for <ed25519.curve.n?

Asking here so we can follow that in our deterministic signature based keygen.

Comment (paulmillr) [reply src/ipns.ts:44]: No need: seed is passed to `getPublicKey`, which will hash the seed and reduce it modulo n.

The check would have been only required when seed is passed to `ExtendedPoint.BASE.multiply(seed)`, rare.

**Comment by 0xc0de4c0ffee** [comment]
For future we'll try to explore multi-sig keys and full IPNS records signing features.

+I've learn more from scure-base as we need max optimized base-x encoder/decoder in Solidity contracts for our deeper ENS+IPFS integration. Our current base36 decoder function is ticking ~350k gas *(view only) without overflowing.
Base36 decoder : https://github.com/namesys-eth/ipfs2-eth-resolver/blob/main/src/IPFS2ETH.sol#L166
ENS based IPFS gateway : https://k51qzi5uqu5dhhcu1pop9pynjg2g3l6vrlt379x6huzy2zhyg54o1u6csnuwi3.ipfs2.eth.limo/

**Comment by 0xc0de4c0ffee** [comment]
Err, sorry..
I mixed up pubkey with prefixed pubkeybytes, & then forgot to run test as it was simple one line edit. 🙏 

- [x] https://github.com/namesys-eth/ed25519-keygen/actions/runs/5126458378

**Comment by paulmillr** [comment]
thanks @0xc0de4c0ffee

**Pull request merged**